### ISSUE TITLE
More portable paths for filenames in campaigns and add-ons, like [binary_path]

### ISSUE THREAD

**Issue by stevecotton** [wesnoth/wesnoth#4633]
The directory name used for a campaign or add-on ends up hardcoded in multiple places in it. For example, once in each scenario:
```
01_Rooting_Out_a_Mage.cfg:    map_file=campaigns/Two_Brothers/maps/01_Rooting_Out_a_Mage.map
02_The_Chase.cfg:    map_file=campaigns/Two_Brothers/maps/02_The_Chase.map
etc
```
Every one of those also hardcodes whether it's a built-in campaign or a UMC campaign, which means every AOI scenario will need to change if it moves to UMC.

For images, there's already `[binary_path]`, for other files there should be something that means "the path for this add-on".

**Comment by stevecotton** [comment]
#2180 would be one method, by accepting `../` in relative paths, but this means checking code that allows directory traversal. If we just want to access files inside the same add-on, maybe it can be done by the add-on itself?

For example, in _main.cfg use `#define` to store the value of `{CURRENT_DIRECTORY}`, and then use the new def to access other files.

**Comment by gfgtdf** [comment]
for the simple map_file case we could also.just extand binary_path to.map files (make wesnoth search map_file s in all [binary_patg]/maps


> For example, in _main.cfg use #define to store the value of {CURRENT_DIRECTORY}, and then use the new def to access other files.

This doesn't work directly like that because Marcos inside ma Ros are afaik substituted when the outer.macro.is used and not when they are defined

**Comment by CelticMinstrel** [comment]
I definitely prefer #2180 for this. If someone else wants to implement it in a simpler way, be my guest; otherwise I'll eventually get back to it.

**Comment by stevecotton** [comment]
> for the simple map_file case we could also just extend binary_path to map files (make wesnoth search map_files in all [binary_path]/maps

That's actually really easy to implement. It makes `[binary_path]` a misnomer, as it now also handles a text-based type of file, however that's going to be the correct path for campaigns or add-ons that use the standard layout with images/, maps/, scenarios/, etc.

**Comment by stevecotton** [comment]
Delfador's Memoirs goes for a simpler option, just making a macro for the hardcoded path:
https://github.com/wesnoth/wesnoth/blob/b479d6f7ac53c4fa3c4c2c7d48ab0928bccbd9d7/data/campaigns/Delfadors_Memoirs/utils/dm-macros.cfg#L3-L5

I still think the functionality should be added to `map_file=`, but DM's method is simple and works with stable versions of Wesnoth.

**Comment by CelticMinstrel** [comment]
> That's actually really easy to implement.

Yeah, it's really easy, indeed. I still prefer the `..` method though.

> It makes `[binary_path]` a misnomer, as it now also handles a text-based type of file

Actually it's also already used for `po` files, if I recall correctly. (Maybe `mo` files too, I forget.)

**Comment by gfgtdf** [comment]
> Yeah, it's really easy, indeed. I still prefer the .. method though.

This is about map_file and not macro inclusion, the code that uses map_file does not know the location the wml file that defined the scenario.

**Comment by CelticMinstrel** [comment]
Ah, you do have a point there. I guess it's reasonable to implement both things then.

**Comment by gfgtdf** [comment]
It'd be nice if those new parts would also be supported by wml that's (replace map) etc that take a map_file parameter

**Comment by stevecotton** [comment]
I'm curious about the regex check in replace_map - as it's already limited to files accessible via get_wml_location, what is this check meant to prevent?
https://github.com/wesnoth/wesnoth/blob/8080271231adddb3a3e88297d5fc52d37e3de7d7/src/game_events/action_wml.cpp#L558-L577

**Comment by gfgtdf** [comment]
> I'm curious about the regex check in replace_map

I'm quite sure that I wrote that code but I don't remember why, I was probably just extra cautious. In particular with the existence of wesnoth.read_file nowdays there is little point in these checks.

**Comment by stevecotton** [comment]
@sevu your 8080271231adddb3a3e88297d5fc52d37e3de7d7 (without any changes from this issue) seems to be broken - Endless Night crashes Wesnoth once enough repetitions have happened to cause the `[replace_map]` to fire.

**Comment by sevu** [comment]
@stevecotton thanks. Did the merged PR fix it?

**Comment by stevecotton** [comment]
@sevu yes :)

**Comment by stevecotton** [comment]
Implemented in #4654
